### ISSUE TITLE
pyepics PIP package is installing unwanted shared libraries

### ISSUE THREAD

**Issue by hdante** [pyepics/pyepics#90]
The following files were installed with pyepics, but they shouldn't have been:

  /usr/lib/libCom.so
  /usr/lib/libca.so
 
 Instead, pyepics should use the system provided EPICS libraries.

 Extra: these libraries caused problems when installing other unrelated packages with yum. Example follows attached.

[yum-error-log.txt](https://github.com/pyepics/pyepics/files/1282448/yum-error-log.txt)

**Comment by newville** [comment]
@hdante  Sorry for the trouble.  Yes, with 3.2.7 epics now installs libca and libCom on Linux, Mac, and Windows.  By default. these get installed according to the install method. For the pip install you did, they go to /usr/lib (which makes sense for "/usr/bin/python").   That's how pip install with a typical system-supplied python will work.

Sadly, we decided earlier this year that using "system provided EPICS libraries" was causing too many unresolvable problem.  The default build of libCom.so has the very nasty behavior of linking against libreadline.  Of course, CA does not use readline.  Having libca.so depend on libreadline.so leads to unresolvable conflicts for applications that want to use pyepics but have also linked against a different version of readline (maybe those applications actually use readline).  This conflict does really happen for many folks using a non-system supplied version of Python, such as Anaconda Python.

As a result, we've decided to build and distribute versions of libca.so and libCom.so that are not linked with readline, hoping that these would be suitable defaults for most users of pyepics.   In short the real change was from defaulting to "system-supplied libca" to "not-contaminated-by-readline libca".  You can always specify which version of lib CA to use by setting PYEPICS_LIBCA. 

It seems like the main problem you are seeing is that these versions of libca.so and libCom.so do not work for you, though I don't know why that is.  Is this RHEL6.9?   We've tested the version we distribute on a few different linuxes (including using a docker image with TravisCI), but maybe there is a conflict with RHEL6.9....   I also don't know why yum would complain about these.  Perhaps the files are invalid on RHEL6.9?    Anyone have any thoughts on why that might be?

**Comment by hdante** [comment]
I'm not sure what the messages meant, but it would be reasonable for it not to go into a standard path. It could go to $libdir/pyepics, instead, for example, so that they're isolated only for pyepics use. This is how other libraries store their internal shared libraries. Examples follow:
[henrique.almeida@sol13-linux lib64]$ ls -lR /usr/lib64/openssl/
/usr/lib64/openssl/:
total 4
drwxr-xr-x. 2 root root 4096 Jul 21 16:12 engines

/usr/lib64/openssl/engines:
total 204
-rwxr-xr-x. 1 root root 19368 Jan 30  2017 lib4758cca.so
-rwxr-xr-x. 1 root root 19600 Jan 30  2017 libaep.so
-rwxr-xr-x. 1 root root 15352 Jan 30  2017 libatalla.so
-rwxr-xr-x. 1 root root  6032 Jan 30  2017 libcapi.so
-rwxr-xr-x. 1 root root 24048 Jan 30  2017 libchil.so
-rwxr-xr-x. 1 root root 27832 Jan 30  2017 libcswift.so
-rwxr-xr-x. 1 root root  6024 Jan 30  2017 libgmp.so
-rwxr-xr-x. 1 root root 15256 Jan 30  2017 libnuron.so
-rwxr-xr-x. 1 root root 14664 Jan 30  2017 libpadlock.so
-rwxr-xr-x. 1 root root 23960 Jan 30  2017 libsureware.so
-rwxr-xr-x. 1 root root 23672 Jan 30  2017 libubsec.so

[henrique.almeida@sol13-linux lib64]$ ls -lR /usr/lib64/atlas/
/usr/lib64/atlas/:
total 10276
lrwxrwxrwx. 1 root root      17 Set  6 15:29 libatlas.so.3 -> ./libatlas.so.3.0
-rwxr-xr-x. 1 root root 4570424 Fev  9  2012 libatlas.so.3.0
lrwxrwxrwx. 1 root root      17 Set  6 15:29 libcblas.so.3 -> ./libcblas.so.3.0
-rwxr-xr-x. 1 root root  130872 Fev  9  2012 libcblas.so.3.0
lrwxrwxrwx. 1 root root      19 Set  6 15:29 libclapack.so.3 -> ./libclapack.so.3.0
-rwxr-xr-x. 1 root root   97752 Fev  9  2012 libclapack.so.3.0
lrwxrwxrwx. 1 root root      19 Set  6 15:29 libf77blas.so.3 -> ./libf77blas.so.3.0
-rwxr-xr-x. 1 root root  127888 Fev  9  2012 libf77blas.so.3.0
lrwxrwxrwx. 1 root root      18 Set  6 15:29 liblapack.so.3 -> ./liblapack.so.3.0
-rwxr-xr-x. 1 root root 5326088 Fev  9  2012 liblapack.so.3.0
lrwxrwxrwx. 1 root root      19 Set  6 15:29 libptcblas.so.3 -> ./libptcblas.so.3.0
-rwxr-xr-x. 1 root root  130936 Fev  9  2012 libptcblas.so.3.0
lrwxrwxrwx. 1 root root      21 Set  6 15:29 libptf77blas.so.3 -> ./libptf77blas.so.3.0
-rwxr-xr-x. 1 root root  127952 Fev  9  2012 libptf77blas.so.3.0

etc.

**Comment by newville** [comment]
@hdante  In principle, the libraries could go a lot of places.  In practice, the sanest place to put them is the "standard location for python libraries according to the python setup/install tools".  With `pip install`, you will be installing files in order for that version of Python to use them.  If that's system python on most Linux systems, you'll be writing to /usr.  

When installing from source, `python setup.py install` typically creates one of the dreaded python 'eggs', a zip file that does fully encapsulate all the files including the shared libraries.  Installing with `pip install` gives a different result, putting the python code in a different location and the shared libraries in /usr/lib.  Python setuptools must have some charms, but they are lost on me. ;).

Many people prefer to use per-user Python environments or virtual environments to avoid such conflicts.   Anaconda Python seems especially popular for this use-case.  That is, the typical python approach is not for the packages to change where files are installed relative to the base python directory, but for the user to control the base python directory.

That's not to say we're opposed to changing where libraries get installed, but we have to worry about the differences of Linux, Windows, Mac, and environments like Anaconda.  It would be non-trivial to maintain if we deviate too far from the standard installation tools.

**Comment by dchabot** [comment]
@hdante, this issue was addressed: issue #81 and related PRs #80, #82, and #83.

This should help avoid installation of the pyepics-provided libca and libCom:

```shell
NOLIBCA=1 pip install pyepics
```

**Comment by hdante** [comment]
@dchabot ok, I'll read it later. Thanks for the hint. The only problem I see right now is that you worked around issues with anaconda but broke standard unix setups. Even if it was addressed, the conclusion was a broken setup.

**Comment by newville** [comment]
@hdante We are putting the shared libraries in "the normal place" for a python install.  It is true and sort of weird that what "the normal place" does depend some on how you install, but that is a feature of setuptools, and changing that would be significant work.   Given the use setuptools, I don't see a reason to change how we determine where to put the libraries on installation.

Like @dchabot mentioned, If you want to NOT install the provided libca and libCom, use the documented (http://pyepics.github.io/pyepics/installation.html) method:

    NOLIBCA=1 pip install pyepics

Or, use PYEPICS_LIBCA after installation to specify exactly which libca to use.  If `pip install` with system python breaks your setup because it wants to install to /usr/lib, use a different python environment.

I think we are not solving the wrong problem: Pyepics does not work without a working libca.  For years we provided libca and libCom for Windows, and now we are including them for Mac and Linux too.   The Linux versions have become increasingly important to help users avoid the utterly wrong default behavior on Linux systems of having libca depend on libreadline.  

Also, just to be clear: a fair number of facilities now rely on Anaconda Python. Their default REPL uses libreadline to do normal readline stuff. They *do* provide their own closed environment, including their own libreadline.  And there are conda packages for pyepics.  An unresolvable problem arise when libCom and the Python executable are built with conflicting version of readline.  This conflict is a common occurrence when using Anaconda Python on Linux.  Since libca does not actually ever use readline, any blame for this conflict goes entirely to Epics base: libca should not depend on a library it does not use.  The result is that pyepics cannot reliably use a version of libca built with readline. 

If the concern is that we're installing a file called "libca.so" into some location that some other installation might also write a file called "libca.so", an alternative might be to create our own "libca4py" that does not clash names with other things call "libca.*".  I think no one has investigated
that thoroughly.

Hope that helps you understand why we're not likely to change this approach of providing and installing libca and libCom by default into the "normal place" that Python expects them to be.  We're not saying this approach is perfect, or that we're doing it without bugs,  but it is intentional and solves real problems that many of us actually see.   But, Pyepics cannot rely on system-built libca by default.

**Comment by newville** [comment]
@hdante  
> The only problem I see right now is that you worked around issues with anaconda 
> but broke standard unix setups. Even if it was addressed, the conclusion was a broken setup.

No, we did not break the installation on standard unix setups. The libraries are installed to the normal place that libraries go when installing python packages.

**Comment by hdante** [comment]
@newville You did, because you've overridden the installed libca.so and libCom.so. Since you won't change this, you can close the issue.

**Comment by newville** [comment]
@hdante isn't the very point of installing software to write files on to  your computer's disk in a way that they can be used?  Doing "pip install pyepics" will install files so that pyepics can be used, and it will overwrite old files with the same name.    

Did you have files /usr/lib/libca.so and /usr/lib/libCom.so before installing pyepics?  Do you consider that "standard" --- I don't think I've ever seen Epics libraries installed to /usr/lib/.    For the systems I work with, they are typically in places with names like /usr/local/epics/base-3.15.5/lib/linux-x86_64/libca.so.

But if so, the available solutions include: 

1. use `NOLICA=1 pip install pyepics`.
2. download the tarball and use `python setup.py install`.
3. use a different python environment.

**Comment by hdante** [comment]
@newville if I give you a package that has its custom libssl.so and it overwrites the system's version, so that all security related operations in you ssh, firefox, chrome and e-mail pass through my libssl.so, would you install it in /usr/lib ?

Here goes attached again the log that proves that you've broken my installation.
[yum-error-log.txt](https://github.com/pyepics/pyepics/files/1285355/yum-error-log.txt)

**Comment by newville** [comment]
Uh, we don't install a custom libssl.so, we install libca.so.  I am not aware of a standard unix file with that name. Or do you mean that some other Epics app is also installing a file named libca.so into /usr/lib?   I've never seen that... is that a common location for libca.so?

You have lots of choices of how to install pyepics.  If you want to install it to the system python with `pip install`, that will install files to /usr/lib.   That's how it works.   There are a number of alternatives.

I have no idea why yum is complaining that these files are truncated.   Are they corrupted or something? Even if that were the case, why would corrupted /usr/lib/libca.so break yum?

**Comment by hdante** [comment]
@newville the point is that /usr/lib is the system wide library location and installing it there should be part of the EPICS base installation setup. It gets the same status of libssl.so, that is, another application that does not own it, but uses its own version of it should use a separate directory.

In my current instalation epics base is in /usr/local/epics/lib. In the future, when we create RPM or DEB packages it will go to /usr/lib. It doesn't matter, it's owned by EPICS base. pyepics should have its own copy in its own directory.

I don't know what's the reason for the error message. I will be able to debug it tomorrow. My guess is that yum runs ldconfig or rescans system wide library paths and this library is for a later Red Hat version or has a missing section required by Red Hat.

**Comment by ralphlange** [comment]
On Linux, `/usr/bin` and `/usr/lib` are system locations. All files in those areas are exclusively to be handled by the system's package manager. (This is defined by the FHS [1] and by distribution dependent policies.)
Local installations, i.e. everything not installed as system packages, are supposed to go into the hierarchy below `/usr/local`. This is why existing (local) installations of EPICS are often found there.
On the other hand, EPICS system packages (e.g. the available set of Debian packages [2]) will install their libraries into `/usr/lib`, because they have to (mandated by the FHS and the applicable packaging policy).

Local installs must not write into `/usr/lib`. A local pyepics install overwriting files in `/usr/lib` is effectively breaking a Linux system. (It's probably roughly the equivalent of silently overwriting DLLs in `C:\Windows\SysWOW64`.)
Doing so is also not in the interest of pyepics, because at the next update the system package manager will rightfully overwrite the libraries in `/usr/lib` with the newer version, and pyepics ends up in exactly the situation it wanted to avoid in the first place: using the system's installed libraries that it does not work well with.
A more appropriate private place for pyepics' private libraries from a local install would be below `/usr/local/lib/pythonX.X/dist-packages` or `/usr/local/lib/pyepics`.

[1] https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard
[2] https://github.com/epicsdeb

**Comment by newville** [comment]
@ralphlange @hdante 

> On Linux, /usr/bin and /usr/lib are system locations. All files in those areas are exclusively to be 
> handled by the system's package manager. (This is defined by the FHS [1] and by distribution 
> dependent policies.)

Hm, I do not see exclusivity described in the link you sent.  I don't see any mention of "system package manager" at all.

For "RedHat" (including Fedora and Centos), the package-provided system python is /usr/bin/python (and a hardlink or exact copy at /bin/python).  These are typically symlinked to eventually point int /usr/bin/python2.7 (there are python3.* too).  They also provide, *as a system package*, the "python package manager" /usr/bin/pip (and similarly /bin/pip, both also links to eventually /usr/bin/pip2.7).    

That supplied "package installer for system python", /usr/bin/pip**, will install python packages with PREFIX=/usr.   That is the intended purpose: files will definitely install into /usr/lib/pythonX.Y/site-packages, and may install data file into other places based from that `sys.prefix` of /usr such as /usr/share.  If I understand your argument, it seems that any use of /usr/bin/pip violates the rule of "only system package managers write to /usr/lib and /usr/bin".   OK, then don't use /usr/bin/pip.

If you want to use a different version of Python, that's fine.  I'd even say that it is probably a good idea to not install non-system python packages to the system's python.  RedHat-like systems do include /usr/bin/virtualenv (and for py3!) so that users can create there own environments, including their own version of pip.   But, if you do install Python packages with /usr/bin/pip, it is going to write files under /usr/lib/.

Note that other installation methods (such as `/usr/bin/python setup.py install` from an unpacked source kit of the tarball downloaded from PyPI,) will install the dynamic libraries into a different location, under /usr/lib/pythonX.Y/site-packages, though as an "egg". This will not write a file called /usr/lib/libca.so, but rather something like /usr/lib/pythonX.Y/site-packages/pyepics-3.2.7-pyX.Y.egg/lib/libca.so.  Does that violate the rule of "only package managers write to /usr/lib"?

I think we're open to changing where the default libraries get installed (and, how they are searched for), but that affects many more systems and install methods than just "Redhat-like linuxes".   Perhaps installing to PREFIX/share/pyepics would be used (though, is that "allowed" according to the easy-to-violate-with-system-provided-tools rule)?  

I am definitely not saying the installation scripts are perfect, and there may be mistakes in them.  But,  I sort of think the arguments being made here are really better answered by "don't use system python on RedHat-like linuxes".

**Comment by ralphlange** [comment]
Note that I wrote just about `/usr/bin` and `/usr/lib` and not about the complete directory hierarchies below.
Especially `/usr/lib/<modulename>` is a place where modules have a lot more freedom to install things.

Something that is not `/usr/lib` (or the architecture subdirectories on multi-arch installs) is fine. So, `/usr/lib/pythonX.Y/site-packages/pyepics-.....` is definitely private enough for private libraries.
`/usr/share/...` is more intended for files that are shared between architectures, which binary libraries are definitely not.

**Comment by newville** [comment]
@tacaswell @dchabot  -- any thoughts on what to do about this?    
I'm finding this  (and the discussion at #89) pretty frustrating.   

There are a number of potential solutions to these Issues, but I'm inclined toward "do nothing", and focusing on more interesting and rewarding projects.

**Comment by dchabot** [comment]
> @tacaswell @dchabot -- any thoughts on what to do about this?

IMO if they are going to be installed, 'libca/libCom' should be co-resident with the `epics` package.

eg: on a Debian system this would look like "/usr/local/lib/pythonX.Y/site-packages/epics/lib[ca|Com].so"

`setuptools` supports this (I think?).

**Comment by newville** [comment]
@dchabot Thanks.  Actually, the "it would look like ..... on Debian" and "setuptools supports this (I think?)" are a bit scary, as is the fact that "pip install" and "python setup.py install" give different results.  If nothing else, it suggests that it will be very challenging to support all the variations.

That is, I think that if we try to provide shared libraries for Linux, there will always be people whining that we're doing something wrong, making (absolutely ridiculous) claims that we are breaking their system, or violating some "policy" of their chose distribution.  We can't win. 

So, I propose that we 

1. Install ca/Com libraries on Windows and Mac (for which "system python" is well-isolated in a Framework), but not for Linux. We could make ca/Com libraries for Linux available as a separate download and pointing to them in the documentation, but not during install.

2. Alter the "find_libca()" mechanism to favor PYEPICS_LIBCA, then look within the local install (which will work for Windows and with Anaconda packages, see below), then search PATH and LD_LIBRARY_PATH, and then give up.  I think we do not need to search for libCom.

3. We make Anaconda packages that include the ca/Com shared libraries.

I do not plan to maintain packages for other package managers (debs, rpms, etc) myself, but it would be great if someone else wanted to do that.

**Comment by ralphlange** [comment]
Complaining about an application installation process silently replacing system libraries is IMHO not ridiculous, as that is a behavior pattern often found with malware.

Putting the pyepics-supplied libraries in the same location that pyepics is installed (like @dchabot suggests) will work across all Linux distributions, and also ensure that parallel installs of multiple pyepics versions will each work with their supplied library.

**Comment by newville** [comment]
@ralphlange  PR welcome.

**Comment by dchabot** [comment]
I'm happy to take a stab at a PR for this...

**Comment by newville** [comment]
@dchabot OK, go for it. I'm not sure I see an easy way to use setuptools and both `python setup.py install` and `pip install` and work consistently across platforms without real effort.    We tried several months back, and this is the solution we came up with.  

It is unfortunate that it doesn't work ideally for the very-much-not-recommended use of `sudo pip install` with a system-provided Python.  Then again, comments from non-contributors that installing pyepics has broken someones system or describe it as malware are also unfortunate and unhelpful. All work done for this project is done out of kindness and an attempt to help others, and all the work done here is a gift we are giving to anyone who wants it.  We are not obligated to place any weight to people who complain without helping, and the more hyperbolic the complaints the less credible they are and the less inclined we *should* be to help them.  The first goal is that it works for us, the people who do the work.  If it is useful to others, great.  If not, that's OK too.  

Then again, we should probably fix the problem of whether (and if so, how) to best install shared libraries on linux.  What that fix looks like is up to us. 

It may be that `package_data` is needed, but I think that *requires* moving the library data files from the current place  to under the 'epics' folder at install time (perhaps copied within setup.py, just prior to running `setup()`?).   That seems kind of fragile, but it might be OK.

I might suggest instead creating a PREFIX/share/pyepics lib and forcing installs of the libraries to this location *outside* the `setup()` command.   I think that might be more reliable than using anything from within setup() and avoid the inevitable pain of using setuptools.
